interpreter/cling/include/cling/Interpreter/Value.h

```diff
@@ -284,6 +250,15 @@ namespace cling {
     void print(llvm::raw_ostream& Out, bool escape = false) const;
     void dump(bool escape = true) const;
   };
+
+  template <> void* Value::getAs() const;
+#define X(type, name)                                                   \
+  template <> type Value::getAs() const;                                \
```

Comment:
How are these explicit instantiation request help?

Reply:
This allows us to generate the relevant code which does the right dispatch. Take a look at the source file. Or I am missing something?

Reply:
I missed that those are intended to a specialization forward declaration rather than instantiation request.  That said, there does seem to be any reason to specialization `getAs`, the code in the macro seems to be generic, isn't it? (On the other `Create` seems to need them)

Reply:
The reason for the specialization of `getAs` is to expose to the compiler the expected return type so that it can make the relevant conversions. The approach is quite similar to what we had in the `ValHolder` in TClingCallFunc. I believe this capture the intent https://godbolt.org/z/E6q7vEM6f 

There you can take a look in the generated code:
  * `getAs` enumerates all types we explicitly model in the union m_Storage so that conversions are safe as the compiler will see both types and generate the right code for conversions;
  * `convert` takes the concrete underlying type (in the example `m_Kind` but in this PR the `clang::QualType`) and makes sure it calls the right accessor of the union m_Storage.

The implementation could have been simpler if we could include `clang/AST/Type.h` in `Value.h`. However, `Value.h` in being included in the interpreter runtime so we should not parse that much code, hence the out-of-line explicit specializations.

Reply:
Yes, exactly.